From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: no name <git-am@invalid>
Date: Sun, 9 Aug 2026 14:00:00 -0600
Subject: [PATCH] multi: transport frame-synthesis swapchains

Forward color, motion-vector, and depth swapchains through the Vulkan and
OpenGL client wrappers. Keep all three references while a frame passes
through Monado's multi-client compositor, then submit the complete
frame-synthesis layer to the native compositor.
---
 src/xrt/compositor/client/comp_gl_client.c       | 26 ++++++++++++++++
 src/xrt/compositor/client/comp_vk_client.c       | 24 +++++++++++++++
 src/xrt/compositor/multi/comp_multi_compositor.c | 24 +++++++++++++++++
 src/xrt/compositor/multi/comp_multi_private.h    |  2 +-
 src/xrt/compositor/multi/comp_multi_system.c     | 39 +++++++++++++++++++++++-
 5 files changed, 113 insertions(+), 2 deletions(-)

diff --git a/src/xrt/compositor/client/comp_gl_client.c b/src/xrt/compositor/client/comp_gl_client.c
index 550c80319..8e9eb1f16 100644
--- a/src/xrt/compositor/client/comp_gl_client.c
+++ b/src/xrt/compositor/client/comp_gl_client.c
@@ -272,6 +272,31 @@ client_gl_compositor_layer_projection_depth(struct xrt_compositor *xc,
 	return xrt_comp_layer_projection_depth(xcn, xdev, xscn, d_xscn, &d);
 }
 
+static xrt_result_t
+client_gl_compositor_layer_projection_frame_synthesis(struct xrt_compositor *xc,
+                                                      struct xrt_device *xdev,
+                                                      struct xrt_swapchain *xsc[XRT_MAX_VIEWS],
+                                                      struct xrt_swapchain *mv_xsc[XRT_MAX_VIEWS],
+                                                      struct xrt_swapchain *d_xsc[XRT_MAX_VIEWS],
+                                                      const struct xrt_layer_data *data)
+{
+	struct xrt_compositor *xcn = to_native_compositor(xc);
+	struct xrt_swapchain *xscn[XRT_MAX_VIEWS];
+	struct xrt_swapchain *mv_xscn[XRT_MAX_VIEWS];
+	struct xrt_swapchain *d_xscn[XRT_MAX_VIEWS];
+
+	assert(data->type == XRT_LAYER_PROJECTION);
+	for (uint32_t i = 0; i < data->view_count; ++i) {
+		xscn[i] = to_native_swapchain(xsc[i]);
+		mv_xscn[i] = to_native_swapchain(mv_xsc[i]);
+		d_xscn[i] = to_native_swapchain(d_xsc[i]);
+	}
+
+	struct xrt_layer_data d = *data;
+	d.flip_y = !d.flip_y;
+	return xrt_comp_layer_projection_frame_synthesis(xcn, xdev, xscn, mv_xscn, d_xscn, &d);
+}
+
 static xrt_result_t
 client_gl_compositor_layer_quad(struct xrt_compositor *xc,
                                 struct xrt_device *xdev,
@@ -607,6 +632,7 @@ client_gl_compositor_init(struct client_gl_compositor *c,
 	c->base.base.layer_begin = client_gl_compositor_layer_begin;
 	c->base.base.layer_projection = client_gl_compositor_layer_projection;
 	c->base.base.layer_projection_depth = client_gl_compositor_layer_projection_depth;
+	c->base.base.layer_projection_frame_synthesis = client_gl_compositor_layer_projection_frame_synthesis;
 	c->base.base.layer_quad = client_gl_compositor_layer_quad;
 	c->base.base.layer_cube = client_gl_compositor_layer_cube;
 	c->base.base.layer_cylinder = client_gl_compositor_layer_cylinder;
diff --git a/src/xrt/compositor/client/comp_vk_client.c b/src/xrt/compositor/client/comp_vk_client.c
index 45192a68f..a01cc11d7 100644
--- a/src/xrt/compositor/client/comp_vk_client.c
+++ b/src/xrt/compositor/client/comp_vk_client.c
@@ -512,6 +512,29 @@ client_vk_compositor_layer_stereo_projection_depth(struct xrt_compositor *xc,
 	return xrt_comp_layer_projection_depth(xcn, xdev, xscn, d_xscn, data);
 }
 
+static xrt_result_t
+client_vk_compositor_layer_projection_frame_synthesis(struct xrt_compositor *xc,
+                                                      struct xrt_device *xdev,
+                                                      struct xrt_swapchain *xsc[XRT_MAX_VIEWS],
+                                                      struct xrt_swapchain *mv_xsc[XRT_MAX_VIEWS],
+                                                      struct xrt_swapchain *d_xsc[XRT_MAX_VIEWS],
+                                                      const struct xrt_layer_data *data)
+{
+	struct xrt_compositor *xcn = to_native_compositor(xc);
+	struct xrt_swapchain *xscn[XRT_MAX_VIEWS];
+	struct xrt_swapchain *mv_xscn[XRT_MAX_VIEWS];
+	struct xrt_swapchain *d_xscn[XRT_MAX_VIEWS];
+
+	assert(data->type == XRT_LAYER_PROJECTION);
+	for (uint32_t i = 0; i < data->view_count; ++i) {
+		xscn[i] = to_native_swapchain(xsc[i]);
+		mv_xscn[i] = to_native_swapchain(mv_xsc[i]);
+		d_xscn[i] = to_native_swapchain(d_xsc[i]);
+	}
+
+	return xrt_comp_layer_projection_frame_synthesis(xcn, xdev, xscn, mv_xscn, d_xscn, data);
+}
+
 static xrt_result_t
 client_vk_compositor_layer_quad(struct xrt_compositor *xc,
                                 struct xrt_device *xdev,
@@ -837,6 +860,7 @@ client_vk_compositor_create(struct xrt_compositor_native *xcn,
 	c->base.base.layer_begin = client_vk_compositor_layer_begin;
 	c->base.base.layer_projection = client_vk_compositor_layer_projection;
 	c->base.base.layer_projection_depth = client_vk_compositor_layer_stereo_projection_depth;
+	c->base.base.layer_projection_frame_synthesis = client_vk_compositor_layer_projection_frame_synthesis;
 	c->base.base.layer_quad = client_vk_compositor_layer_quad;
 	c->base.base.layer_cube = client_vk_compositor_layer_cube;
 	c->base.base.layer_cylinder = client_vk_compositor_layer_cylinder;

diff --git a/src/xrt/compositor/multi/comp_multi_compositor.c b/src/xrt/compositor/multi/comp_multi_compositor.c
index 71c5ebf9e..57f45eab8 100644
--- a/src/xrt/compositor/multi/comp_multi_compositor.c
+++ b/src/xrt/compositor/multi/comp_multi_compositor.c
@@ -708,6 +708,29 @@ multi_compositor_layer_projection_depth(struct xrt_compositor *xc,
 	return XRT_SUCCESS;
 }
 
+static xrt_result_t
+multi_compositor_layer_projection_frame_synthesis(struct xrt_compositor *xc,
+                                                  struct xrt_device *xdev,
+                                                  struct xrt_swapchain *xsc[XRT_MAX_VIEWS],
+                                                  struct xrt_swapchain *mv_xsc[XRT_MAX_VIEWS],
+                                                  struct xrt_swapchain *d_xsc[XRT_MAX_VIEWS],
+                                                  const struct xrt_layer_data *data)
+{
+	struct multi_compositor *mc = multi_compositor(xc);
+
+	size_t index = mc->progress.layer_count++;
+	mc->progress.layers[index].xdev = xdev;
+
+	for (uint32_t i = 0; i < data->view_count; ++i) {
+		xrt_swapchain_reference(&mc->progress.layers[index].xscs[i], xsc[i]);
+		xrt_swapchain_reference(&mc->progress.layers[index].xscs[XRT_MAX_VIEWS + i], mv_xsc[i]);
+		xrt_swapchain_reference(&mc->progress.layers[index].xscs[2 * XRT_MAX_VIEWS + i], d_xsc[i]);
+	}
+	mc->progress.layers[index].data = *data;
+
+	return XRT_SUCCESS;
+}
+
 static xrt_result_t
 multi_compositor_layer_quad(struct xrt_compositor *xc,
                             struct xrt_device *xdev,
@@ -1009,6 +1032,7 @@ multi_compositor_create(struct multi_system_compositor *msc,
 	mc->base.base.layer_begin = multi_compositor_layer_begin;
 	mc->base.base.layer_projection = multi_compositor_layer_projection;
 	mc->base.base.layer_projection_depth = multi_compositor_layer_projection_depth;
+	mc->base.base.layer_projection_frame_synthesis = multi_compositor_layer_projection_frame_synthesis;
 	mc->base.base.layer_quad = multi_compositor_layer_quad;
 	mc->base.base.layer_cube = multi_compositor_layer_cube;
 	mc->base.base.layer_cylinder = multi_compositor_layer_cylinder;
diff --git a/src/xrt/compositor/multi/comp_multi_private.h b/src/xrt/compositor/multi/comp_multi_private.h
index 588878e38..f4ce2309e 100644
--- a/src/xrt/compositor/multi/comp_multi_private.h
+++ b/src/xrt/compositor/multi/comp_multi_private.h
@@ -69,7 +69,7 @@ struct multi_layer_entry
 	 *
 	 * How many are actually used depends on the value of @p data.type
 	 */
-	struct xrt_swapchain *xscs[2 * XRT_MAX_VIEWS];
+	struct xrt_swapchain *xscs[3 * XRT_MAX_VIEWS];
 
 	/*!
 	 * All basic (trivially-serializable) data associated with a layer,
diff --git a/src/xrt/compositor/multi/comp_multi_system.c b/src/xrt/compositor/multi/comp_multi_system.c
index c0c6c9f60..79429d44b 100644
--- a/src/xrt/compositor/multi/comp_multi_system.c
+++ b/src/xrt/compositor/multi/comp_multi_system.c
@@ -105,6 +105,37 @@ do_projection_layer_depth(struct xrt_compositor *xc,
 	xrt_comp_layer_projection_depth(xc, xdev, xsc, d_xsc, data);
 }
 
+static void
+do_projection_layer_frame_synthesis(struct xrt_compositor *xc,
+                                    struct multi_compositor *mc,
+                                    struct multi_layer_entry *layer,
+                                    uint32_t i)
+{
+	struct xrt_device *xdev = layer->xdev;
+	struct xrt_swapchain *xsc[XRT_MAX_VIEWS];
+	struct xrt_swapchain *mv_xsc[XRT_MAX_VIEWS];
+	struct xrt_swapchain *d_xsc[XRT_MAX_VIEWS];
+	struct xrt_layer_data *data = (struct xrt_layer_data *)&layer->data;
+
+	for (uint32_t j = 0; j < data->view_count; j++) {
+		xsc[j] = layer->xscs[j];
+		mv_xsc[j] = layer->xscs[XRT_MAX_VIEWS + j];
+		d_xsc[j] = layer->xscs[2 * XRT_MAX_VIEWS + j];
+
+		if (xsc[j] == NULL || mv_xsc[j] == NULL || d_xsc[j] == NULL) {
+			U_LOG_E("Invalid swap chain for frame-synthesis projection layer #%u!", i);
+			return;
+		}
+	}
+
+	if (xdev == NULL) {
+		U_LOG_E("Invalid xdev for frame-synthesis projection layer #%u!", i);
+		return;
+	}
+
+	xrt_comp_layer_projection_frame_synthesis(xc, xdev, xsc, mv_xsc, d_xsc, data);
+}
+
 static bool
 do_single(struct xrt_compositor *xc,
           struct multi_compositor *mc,
@@ -367,7 +398,13 @@ transfer_layers_locked(struct multi_system_compositor *msc, int64_t display_time
 			struct multi_layer_entry *layer = &mc->delivered.layers[i];
 
 			switch (layer->data.type) {
-			case XRT_LAYER_PROJECTION: do_projection_layer(xc, mc, layer, i); break;
+			case XRT_LAYER_PROJECTION:
+				if (layer->data.view_count > 0 && layer->data.proj.v[0].frame_synthesis.valid) {
+					do_projection_layer_frame_synthesis(xc, mc, layer, i);
+				} else {
+					do_projection_layer(xc, mc, layer, i);
+				}
+				break;
 			case XRT_LAYER_PROJECTION_DEPTH: do_projection_layer_depth(xc, mc, layer, i); break;
 			case XRT_LAYER_QUAD: do_quad_layer(xc, mc, layer, i); break;
 			case XRT_LAYER_CUBE: do_cube_layer(xc, mc, layer, i); break;
-- 
2.51.0
